### PR TITLE
Add tile/tilesheet/tilemap tests 

### PR DESCRIPTION
**Pull request by txanatan**
Opening PR for comment. Will merge when it's ready (pending a once-over and possibly some cleanup)

### CODE REVIEW

**Comment by txanatan** [comment]
Ready to merge, everything looks good to me.

File: tests/test_tilemap.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                                            "tilesheet.png"))
+
+        tilesheet = sappho.Tilesheet.from_file(path, 1, 1)
+
+        # Grab the tile at (0, 0) and blit it's subsurface to another surface,
+        # then compare it against a master surface to ensure it's the color we
+        # want
+
+        target_surface = pygame.surface.Surface((1, 1))
+        target_surface.blit(tilesheet.tiles[0].image, (0, 0))
```
Comment (lillian-lemmer) [inline tests/test_tilemap.py]:
```subsurface = tilesheet.tiles[0].image
target_surface = pygame.surface.Surface((1, 1))
target_surface.blit(subsurface, (0, 0))
```
would be clearer

File: tests/test_tilemap.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        master_surface.fill((255, 0, 0))
+
+        target_view = target_surface.get_view().raw
+        master_view = master_surface.get_view().raw
+        
+        assert(target_view == master_view)
+
+
+class TestTilemap(object):
+    TILEMAP_CSV = """\
```
Comment (lillian-lemmer) [inline tests/test_tilemap.py]:
```
TILEMAP_CSV = """\
              0,1,2
              5,3,4
              """
```

**Comment by lillian-lemmer** [comment]
tests failing on my system. will report more.

File: tests/test_tilemap.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    0,1,2
+    5,3,4
+    """
+    
+    def setup(self):
+        # Load a tilesheet to use for our tilemap
+
```
Comment (lillian-lemmer) [inline tests/test_tilemap.py]: nit: if a comment is important enough to merit a blank line beneath it, it probably should be a docstring

File: tests/test_tilemap.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        testpath = os.path.realpath(__file__)
+        path = os.path.abspath(os.path.join(testpath,
+                                            "..",
+                                            "resources",
+                                            "tilesheet.png"))
+
+        self.tilesheet = sappho.Tilesheet.from_file(path, 1, 1)
+
+    def test_from_csv(self):
+        csv = textwrap.dedent(self.TILEMAP_CSV).strip()
```
Comment (lillian-lemmer) [inline tests/test_tilemap.py]: Again, why the escape character in the constant if we're stripping here? Doesn't strip remove that?

File: tests/test_tilemap.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        assert(len(tilemap.solid_blocks) == 1)
+        assert(tilemap.solid_blocks[0].topleft == (0, 0))
+        assert(tilemap.solid_blocks[0].bottomright == (1, 1))
+
+    def test_render(self):
+        csv = textwrap.dedent(self.TILEMAP_CSV).strip()
+        tilemap = sappho.TileMap.from_csv_string_and_tilesheet(csv,
+                                                               self.tilesheet)
+
+        # Create a surface that has 1x2 strips of red, green, and blue to compare
```
Comment (lillian-lemmer) [inline tests/test_tilemap.py]: this is good but it'd be better in a docstring.

**Comment by lillian-lemmer** [comment]
In general explain high level logic in docstrings, with only helper comments for inline

**Comment by txanatan** [comment]
@lillian-lemmer I've fixed those tests hopefully, would appreciate you testing them on your system to make sure I haven't missed anything

**Comment by lillian-lemmer** [comment]
```
➜  sappho git:(master) ✗ git merge --no-ff improve-test-coverage
Merge made by the 'recursive' strategy.
 tests/resources/tilemap.tmx         |  12 ++++
 tests/resources/tilesheet.png       | Bin 0 -> 80 bytes
 tests/resources/tilesheet.png.rules |   1 +
 tests/test_tilemap.py               | 120 ++++++++++++++++++++++++++++++++++++
 4 files changed, 133 insertions(+)
 create mode 100644 tests/resources/tilemap.tmx
 create mode 100644 tests/resources/tilesheet.png
 create mode 100644 tests/resources/tilesheet.png.rules
 create mode 100644 tests/test_tilemap.py
➜  sappho git:(master) ✗ py.test tests
============================= test session starts ==============================
platform darwin -- Python 2.7.10, pytest-2.9.0, py-1.4.31, pluggy-0.3.1
rootdir: /Users/lillianlemmer/repos/sappho, inifile:
plugins: cov-2.2.1, pep8-1.0.6, xdist-1.14
collected 13 items

tests/test_animatedsprite.py ..
tests/test_layers.py .....
tests/test_tilemap.py ..F..F

=================================== FAILURES ===================================
________________________ TestTilesheet.test_subsurface _________________________

self = <sappho.tests.test_tilemap.TestTilesheet object at 0x101e1edd0>

    def test_subsurface(self):
        testpath = os.path.realpath(__file__)
        path = os.path.abspath(os.path.join(testpath,
                                            "..",
                                            "resources",
                                            "tilesheet.png"))

        tilesheet = sappho.Tilesheet.from_file(path, 1, 1)

        # Grab the tile at (0, 0) and blit it's subsurface to another surface,
        # then compare it against a master surface to ensure it's the color we
        # want

        target_surface = pygame.surface.Surface((1, 1))
        target_surface.blit(tilesheet.tiles[0].image, (0, 0))

        master_surface = pygame.surface.Surface((1, 1))
        master_surface.fill((255, 0, 0))

        target_view = target_surface.get_view().raw
        master_view = master_surface.get_view().raw

>       assert(target_view == master_view)
E       assert '\x00\x00\x00\xff' == '\x00\x00\xff\x00'
E         - \x00\x00\x00\xff
E         ? ----
E         + \x00\x00\xff\x00
E         ?             ++++

tests/test_tilemap.py:54: AssertionError
___________________________ TestTilemap.test_render ____________________________

self = <sappho.tests.test_tilemap.TestTilemap object at 0x101e4eb10>

    def test_render(self):
        csv = textwrap.dedent(self.TILEMAP_CSV).strip()
        tilemap = sappho.TileMap.from_csv_string_and_tilesheet(csv,
                                                               self.tilesheet)

        # Create a surface that has 1x2 strips of red, green, and blue to compare
        # against the rendered tilemap. This surface has to have the SRCALPHA flag
        # and a depth of 32 to match the surface returned by the render function.
        test_surface = pygame.surface.Surface((3, 2), pygame.SRCALPHA, 32)
        test_surface.fill((255, 0, 0), pygame.Rect(0, 0, 1, 2))
        test_surface.fill((0, 255, 0), pygame.Rect(1, 0, 1, 2))
        test_surface.fill((0, 0, 255), pygame.Rect(2, 0, 1, 2))

        # Render the tilemap
        output_surface = tilemap.to_surface()

        # Compare the two surfaces
        test_view = test_surface.get_view().raw
        output_view = output_surface.get_view().raw
>       assert(test_view == output_view)
E       assert '\x00\x00\xff...f\x00\x00\xff' == '\x00\x00\x00\...0\x00\x00\xff'
E         - \x00\x00\xff\xff\x00\xff\x00\xff\xff\x00\x00\xff\x00\x00\xff\xff\x00\xff\x00\xff\xff\x00\x00\xff
E         + \x00\x00\x00\xff\x00\x00\x00\xff\x00\xa0\x00\xff\xd4\x00\x00\xff\x00\x01\xa6\xff\x00\x00\x00\xff

tests/test_tilemap.py:120: AssertionError
===================== 2 failed, 11 passed in 0.96 seconds ======================
```

**Comment by lillian-lemmer** [comment]
I've merged/fixed this. :)

**Comment by lillian-lemmer** [comment]
Uh oh, we should just have a convo about this. I replaced your `common.compare_pygame_surfaces`.

**Comment by lillian-lemmer** [comment]
WHERE TO GO FROM HERE: Please close this pull request, and rebase from master.

**Comment by lillian-lemmer** [comment]
I fixed this myself! Good work!

**Pull request closed without merging**